python/ray/data/dataset.py

```
@@ -2868,75 +2832,72 @@ def to_tf(
+        def validate_column(column: str) -> None:
+            valid_columns = self.schema().names
+            if column not in valid_columns:
+                raise ValueError(
+                    f"You specified '{column}' in `feature_columns` or "
+                    f"`label_columns`, but there's no column named '{column}' in the "
+                    f"dataset. Valid column names are: {valid_columns}."
+                )
 
-        def make_generator():
-            for batch in self.iter_batches(
+        def validate_columns(columns: Union[str, List]) -> None:
+            if isinstance(columns, list):
+                for column in columns:
+                    validate_column(column)
+            else:
+                validate_column(columns)
+
+        validate_columns(feature_columns)
+        validate_columns(label_columns)
+
+        def get_columns_from_batch(
+            batch: Dict[str, tf.Tensor], *, columns: Union[str, List[str]]
+        ) -> Union[tf.Tensor, Dict[str, tf.Tensor]]:
+            assert isinstance(columns, (str, list))
+            if isinstance(columns, str):
+                return batch[columns]
+            return {column: batch[column] for column in columns}
+
+        def generator():
+            for batch in self.iter_tf_batches(
                 prefetch_blocks=prefetch_blocks,
                 batch_size=batch_size,
-                batch_format="pandas",
                 drop_last=drop_last,
                 local_shuffle_buffer_size=local_shuffle_buffer_size,
                 local_shuffle_seed=local_shuffle_seed,
             ):
-                if label_column:
-                    targets = convert_pandas_to_tf_tensor(batch[[label_column]])
-                    if targets.ndim == 2 and targets.shape[1] == 1:
-                        targets = tf.squeeze(targets, axis=1)
-                    batch.pop(label_column)
+                assert isinstance(batch, dict)
```

This won't be the case for tensor datasets (e.g. datasets created via `read_images()` or `from_numpy()`), since these will be single TensorFlow tensors rather than dicts of tensors. I know that isn't supported by the current API since `label_column` is required, but we might want to raise a better error than an `AssertionError` if the user calls this API on such a dataset.